src/api/app/models/cloud/upload_job.rb

```diff
@@ -4,32 +4,51 @@ class UploadJob
     include ActiveModel::Model
     extend Forwardable
 
-    attr_accessor :user_upload_job, :backend_upload_job
-    validate :validate_jobs
+    attr_accessor :user_upload_job, :backend_upload_job, :target_validator, :filename, :architecture, :target, :user
+    validate :validate_dependencies
+    validates :user, presence: true
+    validates :filename, presence: true, format: {
+      with: /\A.+(.raw.xz|.vhdfixed.xz)\z/, message: "'%{value}' is not a valid cloud image (needs to be a raw.xz or vhdfixed.xz file)"
+    }
+    validates :architecture, inclusion: { in: ['x86_64'], message: "'%{value}' is not a valid cloud architecture" }
+    validates :target, inclusion: { in: ['ec2'] }
+
     def_delegator :backend_upload_job, :id
 
-    def self.create(user, params)
-      upload_job = new
-      upload_job.backend_upload_job = Cloud::Backend::UploadJob.create(user, params)
-      return upload_job unless upload_job.backend_upload_job.valid?
+    def self.create(params)
+      upload_job = new(
+        filename: params[:filename],
+        architecture: params[:arch],
+        user: params[:user],
+        target: params[:target]
+      )
+      return upload_job if upload_job.invalid?
+
+      upload_job.target_validator = "::Cloud::Params::#{upload_job.target.capitalize}".constantize.build(params)
```

Comment:
What about making this an instance method? Then you can get rid of this line. i.e.
```ruby
def target_validator
  @target_validator ||= "::Cloud::Params::#{target.capitalize}".constantize.build(params)
end
```

Reply:
Nice idea!